text/0000-dynamic-events.md

```diff
@@ -0,0 +1,146 @@
+---
+title: `lwc:on` directive 
+status: DRAFTED
+created_at: 2025-01-13
+updated_at: 2025-01-13
+pr: (leave this empty until the PR is created)
+---
+
+# `lwc:on` Directive
+
+## Summary
+
+This proposal adds a mechanism to add a collection of event listeners to elements in an LWC template using a new directive `lwc:on`.
+
+## Basic example
+
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+
+    childEventHandlers = {
+        foo: function (event) {console.log('foo');} ,
+        bar: function (event) {console.log('bar');}
+    };
+
+}
+```
+
+```html
+ <!-- x/myComponent.html -->
+<template>
+  <x-child lwc:on={childEventHandlers}></x-child>
+</template>
+```
+
+For each property in `childEventHandlers`, the new `lwc:on` directive adds an event listener to `x-child`. The event listener listens for the event specified by the property's key and uses the property's value, bound to the instance of `MyComponent`, as the event handler.
+
+## Motivation
+
+LWC's support for declarative event listeners is limited to scenarios where event name are known while authoring the owner component template. This is sufficient for most of the cases, however when using the `lwc:component` directive, it is common to require event listeners based on the current value of the constructor passed to `lwc:is`. In such cases, it is not possible to know the event name while authoring the owner component template.
+
+As an alternative, it is possible to add event listeners imperatively. However it requires the owner component to have a reference of the element on which it needs to add event listener. The owner component can have a reference of a element only after the element has been connected and hence typically only after the connectedCallback of corresponding component has terminated , making it impossible for the owner component to handle events dispatched by the connectedCallback of the corresponding component.
+
+
+## Detailed design
+
+A new directive `lwc:on` will be introduced that can be used to add a collection of event listeners whose name may not be known while authoring the component.
+
+### Structure
+
+The `lwc:on` directive would accept an object. For each property of the object, it would add a event listener which would listen for the event specified by the property's key and handle it using the property's value bound to the owner component.
+
+### Caching
+
+#### For static components, i.e. components declared in template directly using its selector
+Since it is uncommon for event listeners to change after the start of Owner component's intial render, We can cache them to improve performance. Note that this is same as how `onevent` on template works currently. For consumers, the implication of this would be that any changes made to the object passed to `lwc:on` after the first `connectedCallback` of owner component would cause no effect.

```

Comment:
```suggestion
Since it is uncommon for event listeners to change after the start of the owner component's intial render, We can cache them to improve performance. Note that this is same as how `onevent` in template HTML works currently. For consumers, the implication of this would be that any changes made to the object passed to `lwc:on` after the first `connectedCallback` of owner component would cause no effect.

```

This makes sense to me. I like keeping the functional behavior the same as `onclick`, `onkeyup`, etc.